### ISSUE TITLE
Questions about data loading

### ISSUE THREAD

**Issue by ahaliassos** [facebookresearch/av_hubert#6]
Hi, excellent work!

I'm trying to understand a couple of details regarding the data loading for the self-supervised pre-training stage. From my understanding (please correct me if I'm wrong), in the default setting, each mini-batch contains sequences of variable length, potentially ranging from `max_sample_size=5` (corresponding to 0.2 secs at 25fps) to `max_sample_size=500` (20 secs). Is any "uniform batching" performed such that each batch has sequences of approximately equal length and thus the number of padded zeros are minimised, to speed up training? I couldn't spot it in the code.

Also, in `lrs3_prepare.py`, it seems that utterances longer than 15 secs are trimmed, whereas in `vox_prepare`, no utterances are trimmed. Does this mean that when both LRS3 and VoxCeleb2 are used to pre-train the model, the longest possible LRS3 sequence is 15 secs whereas the longest possible VoxCeleb2 sequence is 20 secs (determined by `max_sample_size`)?

**Comment by chevalierNoir** [comment]
Hi,

1. The sequences will first be sorted according to some order and consecutive sequences will be grouped into batches such that the total number of frames per batch are less than or equal to `dataset.max_tokens` (https://github.com/facebookresearch/av_hubert/blob/main/avhubert/conf/pretrain/base_vox_iter1.yaml#L44), which is done internally by fairseq. We sort the sequences based on their number of frames (https://github.com/facebookresearch/av_hubert/blob/main/avhubert/hubert_dataset.py#L522-L529).
2. For LRS3, an utterance longer than 15s will be split into shorter ones only if it satisfies certain constraints. Specifically, it is split based on the duration of silence (obtained from the alignment provided in LRS3), which is at least 0.4s. The utterances without silence will not be processed and it is also possible that some utterances are still longer than 15s after splitting. Those are still kept, along with long utterances in VoxCeleb2. In training, we then drop them altogether based on `max_keep_sample_size` and `min_keep_sample_size` (https://github.com/facebookresearch/av_hubert/blob/main/avhubert/hubert_dataset.py#L150-L151).

**Comment by ahaliassos** [comment]
Thanks for your response. Just to make sure that my understanding is correct regarding 1: will each epoch see the same exact batches as it saw in the previous one?

**Comment by ahaliassos** [comment]
I think I answered my own question: if each sequence length in the dataset were different, then each epoch would indeed see the same exact batches. But because `np.lexsort` first sorts by length and then by index which by default is random (`shuffle=True`), there will be some randomness in the batches due to some sequences having the same length. Is this correct?

**Comment by chevalierNoir** [comment]
@ahaliassos That's exactly right. The items inside batch will be same if the each sequence is of different length otherwise there will be some randomness when `shuffle=True`.

**Comment by ahaliassos** [comment]
Perfect, thanks a lot!
